modules/r2dbc/src/main/java/org/testcontainers/r2dbc/ConnectionPublisher.java

```diff
@@ -0,0 +1,152 @@
+     */
+    private class StateMachineSubscription implements Subscription {
+
+        private final Subscriber<? super Connection> actual;
+
+        Subscription subscriptionState;
+
+        StateMachineSubscription(Subscriber<? super Connection> actual) {
+            this.actual = actual;
+            subscriptionState = new WaitRequestSubscriptionState();
+        }
+
+        @Override
+        public synchronized void request(long n) {
+            subscriptionState.request(n);
+        }
+
+        @Override
+        public synchronized void cancel() {
+            subscriptionState.cancel();
+        }
+
+        synchronized void transitionTo(SubscriptionState newState) {
+            subscriptionState = newState;
+            newState.enter();
+        }
+
+        abstract class SubscriptionState implements Subscription {
+            void enter() {
+            }
+        }
+
+        class WaitRequestSubscriptionState extends SubscriptionState {
+
+            @Override
+            public void request(long n) {
+                transitionTo(new WaitFutureCompletionSubscriptionState());
+            }
+
+            @Override
+            public void cancel() {
+            }
+        }
+
+        class WaitFutureCompletionSubscriptionState extends SubscriptionState {
+
+            private CompletableFuture<ConnectionFactory> future;
+
+            @Override
+            void enter() {
+                this.future = futureSupplier.get();
+
+                future.whenComplete((connectionFactory, e) -> {
+                    if (e != null) {
+                        actual.onSubscribe(EmptySubscription.INSTANCE);
+                        actual.onError(e);
+                        return;
+                    }
+
+                    Publisher<? extends Connection> publisher = connectionFactory.create();
+                    transitionTo(new ProxySubscriptionState(publisher));
+                });
+            }
+
+            @Override
+            public void request(long n) {
+            }
+
+            @Override
+            public void cancel() {
+                future.cancel(true);
+            }
+        }
+
+        class ProxySubscriptionState extends SubscriptionState implements Subscriber<Connection> {
+
+            private final Publisher<? extends Connection> publisher;
+
+            private Subscription s;
+
+            ProxySubscriptionState(Publisher<? extends Connection> publisher) {
+                this.publisher = publisher;
+            }
+
+            @Override
+            void enter() {
+                publisher.subscribe(this);
+            }
+
+            @Override
+            public void request(long n) {
+                // Ignore
+            }
+
+            @Override
+            public void cancel() {
+                s.cancel();
+            }
+
+            @Override
+            public void onSubscribe(Subscription s) {
+                this.s = s;
+                s.request(1);
```

Comment:
@mp911de does the connection factory absolutely guarantees a `Publisher` with at most one `onNext(Connection)`? if not, @bsideup it might be worth it to consider guarding against flux-like publishers by adding a boolean guard in `onNext` (doesn't even have to be a volatile, since only `onNext` would read/write it)

Reply:
it does (checked with Mark), yes. See the design notes in the Javadoc

Reply:
`connectionFactory` implements single element or failure semantics. Completion without an element or emitting multiple items violates the spec. It's a cold publisher that should allow multiple subscriptions.